src/user_profile.cpp

```diff
@@ -0,0 +1,199 @@
+            unsigned long long unsigned_message_len;
+
+            if (crypto_sign_open(unsigned_message, &unsigned_message_len,
+                signed_message, (int) signed_size, pub_key) != 0) {
+                printf("Keys do not match, will try with another public key\n");
+            }
+
+            else {
+                printf("Keys match!\n");
+                break;
+            }
+        }
+        else {
+            fprintf(stderr, "Error opening udf.signed\n");
+            return -1;
+        }
+    }
+    return 0;
+}
+
+int KeyChecks::sign_file(string udf_file)
+{
+
+    if(initialize() == -1)
+        return -1;
+
+    unsigned char pk[crypto_sign_PUBLICKEYBYTES];
+    unsigned char sk[crypto_sign_SECRETKEYBYTES];
+    unsigned char* signed_message;
+
+    string stored_key = DIR_PATH + "private.key";
+    struct stat buf;
+    if (stat(stored_key.c_str(), &buf) == 0){
+        ifstream key_file(stored_key.c_str(), ios::in|ios::binary|ios::ate);
+
+        if(key_file.is_open()){
+            key_file.seekg (0, ios::beg);
+            key_file.read ((char*)sk, crypto_sign_SECRETKEYBYTES); 
+        }
+
+        crypto_sign_ed25519_sk_to_pk(pk, sk);
+    }
+    else{    
+        crypto_sign_keypair(pk, sk);
+    }
+
+    struct stat statbuf;
+    if (stat(udf_file.c_str(), &statbuf) < 0) {
+        perror(udf_file.c_str());
+        return -1;
+    }
+    char *udf = (char *) calloc(statbuf.st_size, sizeof(char));
+    FILE *fp = fopen(udf_file.c_str(), "r");
+    fread(udf, sizeof(char), statbuf.st_size, fp);
+    fclose(fp);
+        
+    signed_message = new unsigned char[crypto_sign_BYTES + statbuf.st_size];
+    unsigned long long signed_message_len;
+    crypto_sign(signed_message, &signed_message_len,
+        reinterpret_cast<const unsigned char *>(udf), statbuf.st_size, sk);
+
+    Directories d;
+    d.create_dir_struct(pk, sk, signed_message, signed_message_len);
+    
+    return 0;
+}
+
+void Directories::create_dir_struct(unsigned char* pk, unsigned char* sk, 
+    unsigned char* signed_message, unsigned long long signed_message_len)
+{
+    DIR *pDir;
+    pDir = opendir ((DIR_PATH).c_str());
+    if (pDir == NULL)
+    {
+        // Creating a directory 
+        if (mkdir((DIR_PATH).c_str(), 0777) == -1){
+            cerr << "Error: " << strerror(errno) << endl;
+            exit(1);
+        } 
+        if (mkdir((DIR_PATH+"/default").c_str(), 0755) == -1){
+            cerr << "Error: " << strerror(errno) << endl;
+            exit(1);
+        }
+        if (mkdir((DIR_PATH+"/allow").c_str(), 0755) == -1){
+            cerr << "Error: " << strerror(errno) << endl;
+            exit(1);
+        }
+        if (mkdir((DIR_PATH+"/deny").c_str(), 0755) == -1){
+            cerr << "Error: " << strerror(errno) << endl;
+            exit(1);
+        }
+        else
+            printf("Directory created\n");    
+    }
+    (void) closedir (pDir);
+
+    if (1) {
```

Comment:
Since this is a C++ class, you can move the stuff below each `if (1)` to small private methods: `createPublicKey()`, `createPrivateKey()`, and `signFile()`

Reply:
Done